pkg/apis/core/validation/conditional_validation.go

```diff
@@ -40,57 +41,62 @@ func ValidateConditionalService(service, oldService *api.Service, allowedIPFamil
 		}
 	}
 
-	errs = append(errs, validateIPFamily(service, oldService, allowedIPFamilies)...)
+	errs = append(errs, validateIPFamilies(service, oldService, allowedIPFamilies)...)
 
 	return errs
 }
 
-// validateIPFamily checks the IPFamily field.
-func validateIPFamily(service, oldService *api.Service, allowedIPFamilies []api.IPFamily) field.ErrorList {
+// validateIPFamilies checks the IPFamilies field.
+func validateIPFamilies(service, oldService *api.Service, allowedIPFamilies []api.IPFamily) field.ErrorList {
 	var errs field.ErrorList
 
 	// specifically allow an invalid value to remain in storage as long as the user isn't changing it, regardless of gate
-	if oldService != nil && oldService.Spec.IPFamily != nil && service.Spec.IPFamily != nil && *oldService.Spec.IPFamily == *service.Spec.IPFamily {
+	if oldService != nil && oldService.Spec.IPFamilies != nil && service.Spec.IPFamilies != nil && reflect.DeepEqual(oldService.Spec.IPFamilies, service.Spec.IPFamilies) {
```

Comment:
@danwinship do we care of the set between old & new is the same, but just different order? That would be rejected here, but you could argue it's a NOP and should pass validation.

Reply:
it's not a NOP. "dual stack IPv4 primary" and "dual stack IPv6 primary" are different